File: 17/Fair/Activity/fb.js
let fs = require("fs");
let puppeteer = require("puppeteer");
let cFile = process.argv[2];
let pageName = process.argv[3];
let noOfPost = process.argv[4];

(async function(){
    let credentialFile = await fs.promises.readFile(cFile);
    let {user, password, url} = JSON.parse(credentialFile);
    let browser = await puppeteer.launch({
        headless: false,
        defaultViewport: null,
        args: ["--start-maximized", "--disable-notifications"]
    });

    let tabs = await browser.pages();
    let tab = tabs[0];

    //normal => networkidle0
    // front end heavy => networkidle2
    await tab.goto(url, {waitUntil: "networkidle2"});
    await tab.waitForSelector("input[type=email]");
    await tab.type("input[type=email]", user, {delay: 120});
    await tab.type("input[type=password]", password, {delay: 120});

    //click => cause navigation wrap in Promise.all
    await Promise.all([
        tab.click("input[type=submit]", {delay: 120}),
        tab.waitForNavigation({waitUntil: "networkidle2"})
    ])

    // home (page
    // _1frb

    await tab.waitForSelector("._1frb");
    await tab.type("._1frb", pageName, {delay: 120});
    await tab.keyboard.press("Enter");
    await tab.waitForNavigation({waitUntil: "networkidle2"});

    //search page
    await tab.waitForSelector("li[data-edge=keywords_pages]");
    await Promise.all([
        tab.click("li[data-edge=keywords_pages] a"),
        tab.waitForNavigation({waitUntil: "networkidle2"})
    ]);

    await tab.waitForSelector("._1glk._6phc.img");
    await Promise.all([
        tab.waitForNavigation({waitUntil: "networkidle2"}),
        tab.click("._1glk._6phc.img")
    ]);

    //post tab => meaning full selector
    //click
    //data-edge="keywords_pages"

    // url changing twice => so 2 time waitfornavigation()
    await tab.waitForSelector("div[data-key=tab_posts]");
    await Promise.all([
        tab.click("div[data-key=tab_posts] a"),
        tab.waitForNavigation({waitUntil: "networkidle2"})
    ]);
    await tab.waitForNavigation({waitUntil: "networkidle2"});

    // // Method 1 - Like posts
    // // loader after every 7 posts
    // await tab.waitForNavigation({waitUntil: "networkidle2"});

    // for(let i=0; i<noOfPost; i++){
    //     await tab.waitForSelector("._6rk2.img.sp_X73jHqRLgmV.sx_ee9b30");
    //     await Promise.all([
    //         tab.click("._6rk2.img.sp_X73jHqRLgmV.sx_ee9b30")
    //     ]);
    // }

    // Method 2 - Like posts
    // loader after every 7 posts
    // 20 post => 7 before each loader
    let idx = 0;
    do{
        // all posts => request, time => page => element display
        // present => 1xnd => 7 posts
        await tab.waitForSelector(".pagelet_timeline_main_column ._1xnd .clearfix.uiMorePager");
        // get all posts
        let posts = await tab.$$(".pagelet_timeline_main_column ._1xnd>._4-u2._4-u8");
        let cPost = posts[idx];
        let likeBtn = await cPost.$("._666k ._8c74");
        await likeBtn.click({delay: 120});

        //loader => data
        await tab.waitForSelector(".uiMorePagerLoader.pam.uiBoxLightblue", {hidden: true});
        idx++;
    }while(idx<noOfPost);

})();

File: 19/Fair/Activity/fb.rb
require "selenium-webdriver"
include Selenium 
include Selenium::WebDriver

cFile = ARGV[0]
pageToLike = ARGV[1]
numOfPosts = ARGV[2].chomp.to_i

wait = Wait.new(:timeout => 3000)

# options = Selenium::WebDriver::Chrome::Options.new
options = Chrome::Options.new()
options.add_argument("--start-maximized")
options.add_argument("--disable-notifications")
driver = WebDriver.for(:chrome, options: options)
driver.navigate.to "https://facebook.com"
file = File.read(cFile)
map = JSON.parse(file)
user = map["user"]
pwd = map["password"]
# puts pwd
# wait until
wait.until{driver.find_element(:css, "input[type=email]")}
# find
email=driver.find_element(:css,"input[type=email]")
#input
email.send_keys(user)
password = driver.find_element(:css, "input[type=password]")
password.send_keys(pwd)
login = driver.find_element(:css, "input[type=submit]")
login.click()
# driver.quit

#****************Login********************
# wait
wait.until{driver.find_element(:css, "._1frb")}
searchBox = driver.find_element(:css, "._1frb")
searchBox.send_keys(pageToLike)
#searchBox.send_keys(:arrowdown)
searchBox.send_keys(:enter)

#****************search page*****************
wait.until{driver.find_element(:css, "li[data-edge=keywords_pages]")}
pages_tab = driver.find_element(:css, "li[data-edge=keywords_pages] a")
pages_tab.click()
wait.until{driver.find_element(:css,"._1glk._6phc.img")}
pages_img = driver.find_element(:css,"._1glk._6phc.img")
pages_img.click()

wait.until{driver.find_element(:css, "div[data-key=tab_posts]")}
pages_post = driver.find_element(:css, "div[data-key=tab_posts]")
pages_post.click()

idx = 0
while(idx < numOfPosts)
    # all_posts => resuest, time => page -> element display
    # present -> 1xnd => 7 post
    wait.until{driver.until(:css, "#pagelet_timeline_main_column ._1xnd .clearfix.uiMorePager")}
    #get all all_posts

    posts = driver.find_element(:css, "#pagelet_timeline_main_column ._1xnd>._4-u2._4-u8")
    cPost = posts[idx]
    wait.until{driver.find_element(:css, "._666k ._8c74")}
    like_btn = driver.find_element(:css, "._666k ._8c74")
    like_btn.click()
    puts idx
    idx+=1

    #selenium -> wait network => manually scroll
    if(idx == posts.length)
        driver.execute_script("window.scrollTo(0,document.body.clientHeight)")
        sleep 2
    end
end

# for file read - gem install json
# http://www.newthinktank.com/2015/02/ruby-programming-tutorial/
# yahan argv ki jagah ARGV hota hai aur yahan ARGV 0 se start ho jata hai jabki node vala input argv me 2 se start hota tha

# Serial work - using sync programming

# find 
# enter data
